File: frontend/src/services/jobService.js
import api from './api';

export const jobService = {
  getJobs: async () => {
    try {
      const response = await api.get('/jobs');
      return response.data;
    } catch (err) {
      console.warn('Jobs API error or empty, returning default initial jobs:', err);
      return [];
    }
  },

  createJob: async (jobData) => {
    // jobData: { title, description, experience, location, employment_type, salary_min, salary_max, skills }
    const response = await api.post('/jobs', {
      title: jobData.title,
      description: jobData.description,
      experience: Number(jobData.experience || 0),
      location: jobData.location,
      employment_type: jobData.employment_type || 'FULL_TIME',
      salary_min: jobData.salary_min ? Number(jobData.salary_min) : null,
      salary_max: jobData.salary_max ? Number(jobData.salary_max) : null,
      skills: Array.isArray(jobData.skills) ? jobData.skills : jobData.skills.split(',').map((s) => s.trim()).filter(Boolean),
    });
    return response.data;
  },

  getJobByCode: async (code) => {
    const response = await api.get(`/public/jobs/${code}`);
    return response.data;
  },

  getJobById: async (jobId) => {
    const response = await api.get(`/jobs/${jobId}`);
    return response.data;
  },

  updateJob: async (jobId, jobData) => {
    const response = await api.put(`/jobs/${jobId}`, {
      title: jobData.title,
      description: jobData.description,
      experience: Number(jobData.experience || 0),
      location: jobData.location,
      employment_type: jobData.employment_type || 'FULL_TIME',
      salary_min: jobData.salary_min ? Number(jobData.salary_min) : null,
      salary_max: jobData.salary_max ? Number(jobData.salary_max) : null,
      skills: Array.isArray(jobData.skills) ? jobData.skills : jobData.skills.split(',').map((s) => s.trim()).filter(Boolean),
    });
    return response.data;
  },

  deleteJob: async (jobId) => {
    const response = await api.delete(`/jobs/${jobId}`);
    return response.data;
  },
};

export default jobService;

File: backend/app/services/job_service.py
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.job import Job
from app.repositories.job_repository import (
    JobRepository
)

from app.utils.job_utils import (
    generate_application_code
)
class JobService:

    @staticmethod
    async def create_job(
        db: AsyncSession,
        current_user,
        request
    ):

        try:

            job = Job(
                company_id=current_user.company_id,
                title=request.title,
                description=request.description,
                experience=request.experience,
                location=request.location,
                employment_type=request.employment_type,
                salary_min=request.salary_min,
                salary_max=request.salary_max,
                skills=request.skills,
                application_code=
                    generate_application_code()
            )

            job = await (
                JobRepository.create_job(
                    db,
                    job
                )
            )

            await db.commit()

            return job

        except Exception:

            await db.rollback()
            raise

    @staticmethod
    async def get_jobs(
        db,
        current_user
    ):

        return await (
            JobRepository.get_jobs_by_company(
                db,
                current_user.company_id
            )
        )

    @staticmethod
    async def get_job_by_id(
        db: AsyncSession,
        job_id,
        current_user
    ):
        job = await JobRepository.get_by_id(db, job_id)
        if job and job.company_id == current_user.company_id:
            return job
        return None

    @staticmethod
    async def update_job(
        db: AsyncSession,
        job_id,
        request,
        current_user
    ):
        job = await JobRepository.get_by_id(db, job_id)
        if not job or job.company_id != current_user.company_id:
            return None
        
        try:
            job.title = request.title
            job.description = request.description
            job.experience = request.experience
            job.location = request.location
            job.employment_type = request.employment_type
            job.salary_min = request.salary_min
            job.salary_max = request.salary_max
            job.skills = request.skills
            
            await db.commit()
            await db.refresh(job)
            return job
        except Exception:
            await db.rollback()
            raise

    @staticmethod
    async def delete_job(
        db: AsyncSession,
        job_id,
        current_user
    ):
        job = await JobRepository.get_by_id(db, job_id)
        if not job or job.company_id != current_user.company_id:
            return False
        
        try:
            await db.delete(job)
            await db.commit()
            return True
        except Exception:
            await db.rollback()
            raise